Add support for deployments and configmaps in k8s/OS recipes

amisevsk opened the pull request:
### What does this PR do?
Adds support for Deployments and ConfigMaps in Kubernetes and OpenShift recipes.

This PR is split into three commits:
- The first commit adds support for Deployments and makes changes necessary to make main (not test) code compile. This commit only compiles if test compilation is skipped
- The second commit fixes tests to accommodate the changes. It's separated out for readability.
- The third commit adds support for configmaps. I'm lumping it with the rest of the PR because it's closely related and interconnected.

Some relatively major changes here:
- Since Deployments store the underlying pod as `PodTemplateSpec`, there is no way to provide provisioners with the environment's pods directly. As a result, I've added a class internal to `KubernetesEnvironment` that basically abstracts pod spec and template to allow us to return one map of objects to be provisioned. This means relatively minor changes to all the provisioners
   - The new class is `PodData`, and has the same methods as `Pod`s, so that it's mostly a drop-in replacement.
- Since environments now store both pods and/or deployments at the same time, the old methods `getPods()` and it's parallel `getDeployments()` need to be used more carefully. Ideally, they are only used when we need read-only access to these objects (e.g. when we're deploying them). Instead, `getPodData()` should be used, as this returns metadata and specs for all pods/deployments in the environment.
- A lot of tests had to be modified fairly significantly to not accidentally test the scenario where we are returning a modifiable map of pods due to mocking -- ideally there would be a way to assert that provisioners do not modify the pods map directly. In general, this PR makes provisioners more complicated in a way I'm not altogether happy with.
- I had to modify a lot of methods that took pods as parameters to instead take a pod spec and/or pod metadata separately, as there is no pod to pass around for the deployments case.

Feedback on implementation / improvements is more than welcome :)

### What issues does this PR fix or reference?
#11505 and #11507 

#### Release Notes
Add support for Kubernetes deployments and configmaps in Kubernetes and OpenShift recipes

#### Docs PR
Happy to provide if needed.

ibuziuk commented:
ci-test

sleshchenko commented on infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/environment/OpenShiftEnvironment.java:
```diff
@@ -63,13 +66,24 @@ public OpenShiftEnvironment(KubernetesEnvironment k8sEnv) {
       InternalEnvironment internalEnvironment,
       Map<String, Pod> pods,
+      Map<String, Deployment> deployments,
+      Map<String, PodData> podData,
```
I think it's not a good idea to have the ability to have a public constructor where `deployments`, `pods` and `podData` are set by different arguments. As far as I see there are no usages of this constructor but still, Environment created via this constructor may be in inconsistency state.
We are able to avoid even having such protected constructor by moving `podsData` filling logic to `KubernetesEnvironment` constructor from Builders. WDYT?

amisevsk replied:
I think that's a great idea. I'll do that.

riuvshin commented:
Results of automated E2E tests of Eclipse Che Multiuser on OCP:
[Build details](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/1312/)
[Test report](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/1312//Selenium_20tests_20report/)
docker image: eclipseche/che-server:12203
https://github.com/orgs/eclipse/teams/eclipse-che-qa please check this report.

sleshchenko commented:
@ibuziuk `ci-test` job already build PR sources before running tests https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/1312/. But Success Rate is 5% which indicates that something is wrong. What is the purpose of trigger build separately from tests? To have `ci-build-check` passed?

ibuziuk commented:
@sleshchenko I though that both `ci-test` / `ci-build` check are required.

> To have ci-build-check passed

Yeah, as a matter of fact it also fails from time to time, like in the last run with failures in `org.eclipse.che.api.workspace.server.hc.ServerCheckerTest.successfulCheckTest` . Wondering if it is a common problem or related to the changes in this PR ?

sleshchenko commented:
@ibuziuk I've faced the same issue on my PR https://github.com/eclipse/che/pull/12176#issuecomment-447275418 
It's a good time to create an issue for investigating a possible way to make this test stable.

amisevsk commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/Names.java:
```diff
@@ -55,16 +65,16 @@ public static String machineName(Pod pod, Container container) {
+    return podMeta.getName() + '/' + containerName;
   }
 
   /** Return pod name that will be unique for a whole namespace. */
```
Done.

amisevsk commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java:
```diff
@@ -712,39 +718,57 @@ protected void doStartMachine(KubernetesServerResolver serverResolver)
 
-  private void startTracingContainersStartup(Pod pod) {
+  /** Puts createdPod in the {@code machines} map and sends the starting event for this machine */
+  public void startMachines(
```
Good point -- the non private access control was just a mistype by me. Fixed

amisevsk commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesObjectUtil.java:
```diff
@@ -45,6 +46,20 @@ public static void putLabel(HasMetadata target, String key, String value) {
+    putLabel(target.getMetadata(), key, value);
+  }
+
+  public static void putLabel(PodData target, String key, String value) {
```
Good catch -- I must've refactored its use out at some point.

amisevsk commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java:
```diff
@@ -641,12 +645,14 @@ protected void startMachines() throws InfrastructureException {
+      Set<String> toWatch = new HashSet<>();
+      KubernetesEnvironment environment = getContext().getEnvironment();
+      toWatch.addAll(environment.getPodsCopy().keySet());
+      toWatch.addAll(environment.getDeploymentsCopy().keySet());
```
It's tricky, and a complication that this PR adds that I don't really like. As far as I can tell, the pod name specified in a deployment is ignored by OpenShift, and instead the pod just gets the deployment name + hashes. So this does work, it's just confusing. Since we're basically creating the deployment from the yaml directly, it means our recipes are beholden to how OpenShift/Kubernetes handle the objects themselves.

While double checking this, though, I found two issues that are now fixed: Pods weren't getting the deployment name label (accidentally deleted that line in refactoring) and Deployments weren't getting a unique name from the provisioner.

amisevsk commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/KubernetesEnvironment.java:
```diff
@@ -112,9 +135,43 @@ public KubernetesEnvironment setType(String type) {
+   * Add a pod to the current environment. This method is necessary as the map returned by {@link
+   * #getPodsCopy()} is a copy. This method also adds the relevant data to {@link #getPodData()}.
+   */
+  public void addPod(String key, Pod pod) {
```
I agree -- I didn't use `podName` as the parameter name mostly because I wasn't sure if that's always what was used, but it seems to be the case. I'll fix it.

sleshchenko replied:
What about 
```
   public void addPod(Pod pod) {
     String podName = pod.getMetadata().getName;
     pods.put(podName, pod);
     podData.put(podName, new PodData(pod.getSpec(), pod.getMetadata()));
   }
```
?

amisevsk replied:
Good suggestion, that's a bit cleaner. Fixed.

amisevsk commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/KubernetesEnvironment.java:
```diff
@@ -112,9 +135,43 @@ public KubernetesEnvironment setType(String type) {
+   * deployment and pod objects that form the workspace, and should be used when provisioning or
+   * performing any action that needs to see every object in the environment.
+   */
+  public Map<String, PodData> getPodData() {
```
Good point -- I'll rename to `getPodsData`. This makes the method name more in line with `getPodsCopy()` and `getDeploymentsCopy()`. 

As an aside, WDYT about renaming `getPods()` -> `getPodsCopy()`. I did it with the goal of making clear that the map returned cannot be modified, but think the name could be better.

sleshchenko replied:
> As an aside, WDYT about renaming getPods() -> getPodsCopy(). I did it with the goal of making clear that the map returned cannot be modified, but think the name could be better.

I see your point, and it makes sense for me as well. I also think the name could be better but I have no any proposals.

amisevsk commented:
@sleshchenko Thank you for the early review -- I know this is a tedious PR to look over.

I've added a commit to address your comments and fix a few issues I found while double checking it. See the commit message for a detailed list of changes (it changes a method name so of course it touches ~20 files).

Also I forgot to add the recipe I have been using for testing:

```yaml
kind: List
items:
  -
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: ws
      labels:
        test.label: test
    spec:
      replicas: 2
      selector:
        matchLabels:
          che.deployment_test: test
      template:
        metadata:
          name: ws
          labels: 
            test.label: test
        spec:
          containers:
            -
              image: eclipse/che-dev:nightly
              name: dev
              resources:
                limits:
                  memory: 512Mi
              env:
              - 
                name: CONFIGMAP_TEST
                valueFrom:
                  configMapKeyRef:
                    name: test-configmap
                    key: test.entry

  -
    apiVersion: v1
    kind: ConfigMap
    metadata:
      labels:
        test.label: test
      name: test-configmap
    data:
      test.entry: 'test.key'
```
I use this to create a Che 7 workspace based off a deployment, with a configmap included for testing. I basically duplicated the current Che 7 stack and replaced the recipe with this one in my testing.

Of course, functionality for currently existing recipes should be unchanged.

amisevsk commented:
To add: I've been experimenting with how the different metas in a deployment affect the workspace, and I've found:

1. When creating a deployment, the pod template spec meta is *ignored*. This means that no matter what we set the pod name to, it will be overwritten with the deployment name (plus hashes for replica set, etc.)
2. On the Che side, the pod template name is what is used to match with the machine name in the stack, so no matter what the deployment name is, the pod template spec name has to match the machine.

I think we should either 
- Add some docs to explain this discrepancy (I can do that), or 
- Try to fix it (should machine name come from deployment name instead?). This could be very tricky, as `PodData` is used to generate machine names throughout the code. I don't know if this is even possible.

Added a commit to make sure we're getting the machine correctly now. To test, change the `spec.template.metadata.name` in the recipe I posted to something or deployment name so that they don't match (an oversight on my part when writing the recipe).

sleshchenko commented:
@amisevsk Please try to use `PodTemplate#metadata#generateName` instead of `name`. Is it ignored as well? https://docs.openshift.com/container-platform/3.5/rest_api/kubernetes_v1.html#v1-objectmeta

amisevsk commented:
@sleshchenko `generateName` is also ignored, AFAICT. I'm also not sure it would offer many benefits, as it could result in a truncated name (making matching deployed objects to internal machines very difficult). 

I think the purpose of `generateName` is to avoid name collisions when creating multiple objects from the same yaml. The deployment seems to always override it as necessary.

amisevsk commented:
Basically, the only added constraint here is, like in the `pod` recipe case, that the pod name has to match the machine name. The reason it's confusing is that normally it has no effect when writing normal Kubernetes deployments.

I've heard from people that they see stack creation as something of a arcane art, so maybe improving the docs is in order regardless :)

ibuziuk commented:
> I've heard from people that they see stack creation as something of a arcane art, so maybe improving the docs is in order regardless :)

very good point. I do believe we should consider creating a separate issue (probably for the next sprint) which would describe creation of custom stacks based on k8s / openshift  recipes

l0rd commented:
@amisevsk nice job! 

You have probably already planned it but you should update https://www.eclipse.org/che/docs/che-6/recipes.html#kubernetes-yaml-limitations-and-restrictions as well.

Does it work with deployments that have more than one pod?

amisevsk commented:
I've added user dashboard and validation to the PR so that @ibuziuk 's use case (actually creating a new stack with the recipe) should work now. It's my first time doing anything mildly big in the dashboard side so I'm looking forward to some good feedback :)

@l0rd The recipes should have no problem with multiple deployments, but AFAIK Kubernetes deployments support only a single pod spec per deployment. The validation will complain if you ask for more than one replica, and the server-side will overwrite it with "1" even if you manually change the recipe in a new stack. 

As an aside, with the dashboard validation in place, a lot of our discussion about workspace machines is easier to take, as the recipe will correctly (I think) get workspace machines from deployments in recipes. I'm working on a docs PR to update the requirements there.

amisevsk commented:
@l0rd PR updating https://www.eclipse.org/che/docs/che-6/recipes.html#kubernetes-yaml-limitations-and-restrictions is [here](https://github.com/eclipse/che-docs/pull/647)

garagatyi commented:
ci-test

riuvshin commented:
Results of automated E2E tests of Eclipse Che Multiuser on OCP:
[Build details](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/1347/)
[Test report](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/1347//Selenium_20tests_20report/)
docker image: eclipseche/che-server:12203
https://github.com/orgs/eclipse/teams/eclipse-che-qa please check this report.

garagatyi commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/UniqueNamesProvisioner.java:
```diff
@@ -48,15 +59,38 @@ public void provision(T k8sEnv, RuntimeIdentity identity) throws InfrastructureE
+
+    // We expliticly need to modify the deployments in the environment to provision unique names
+    // for them.
+    final Collection<Deployment> deployments = k8sEnv.getDeploymentsCopy().values();
```
I think we need tests to cover all the changes in this class

amisevsk replied:
Good point, will do.

amisevsk replied:
Add a bunch of tests for configmap/deployment renaming, and also for renaming configmaps referenced in pods.

garagatyi commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/KubernetesPluginsToolingApplier.java:
```diff
@@ -84,18 +85,19 @@ public void apply(InternalEnvironment internalEnvironment, Collection<ChePlugin>
     switch (pods.size()) {
       case 0:
         addToolingPod(kubernetesEnvironment);
+        pods = kubernetesEnvironment.getPodsData();
```
Can you elaborate on why we need to reassign the variable here?

amisevsk replied:
`getPodsData()` (and the other methods) have been changed to return immutable copies, so adding the tooling pod won't change `pods`.

garagatyi replied:
ok

amisevsk commented on infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/PerWorkspacePVCStrategyTest.java:
```diff
@@ -108,11 +109,11 @@
   @Mock private EphemeralWorkspaceAdapter ephemeralWorkspaceAdapter;
 
-  private PerWorkspacePVCStrategy perWorkspaceStratefy;
+  private PerWorkspacePVCStrategy perWorkspaceStrategy;
```
From the way it currently works, I don't think so. The only changes to the common/unique strategies is that they have to use `podData` instead of `pod` to catch `Deployments`. This strategy just uses workspace id AFAICT.

garagatyi replied:
ok

amisevsk commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java:
```diff
@@ -712,39 +718,60 @@ protected void doStartMachine(KubernetesServerResolver serverResolver)
+      storeStartingMachine(createdPod, toCreateMeta, machineConfigs, serverResolver);
+    }
+    for (Deployment toCreate : environment.getDeploymentsCopy().values()) {
+      PodTemplateSpec template = toCreate.getSpec().getTemplate();
```
Most of the code added is a refactor to avoid code duplication (since we do basically the same thing for pods and then for deployments.

I'll add tests covering deployments.

amisevsk replied:
Added tests for creating workspace with environment that includes only a deployment and also a deployment and pod.

amisevsk commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/KubernetesEnvironment.java:
```diff
@@ -100,21 +132,72 @@ protected KubernetesEnvironment(
+   * Add a pod to the current environment. This method is necessary as the map returned by {@link
+   * #getPodsCopy()} is a copy. This method also adds the relevant data to {@link #getPodsData()}.
+   */
+  public void addPod(Pod pod) {
```
It shouldn't be a huge issue, and might make the deploy code simpler, but would require rewriting all of the recipes that use pods to use deployments, and modifying the semantics of converting e.g. docker recipes to deployment objects instead of pods. 

Definitely something we should look at though, since this could clean up a fair bit of the deleting workspaces code I think.

amisevsk commented on infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/KubernetesEnvironmentValidatorTest.java:
```diff
@@ -51,7 +52,7 @@ public void setUp() throws Exception {
   public void shouldThrowExceptionWhenEnvDoesNotHaveAnyPods() throws Exception {
     // given
-    when(kubernetesEnvironment.getPods()).thenReturn(emptyMap());
+    when(kubernetesEnvironment.getPodsData()).thenReturn(emptyMap());
```
Sorry, I'm not sure what you mean. Like when `getPodsData()` is not empty but the actual `pods` and `deployments` are? I think this would be a test case on KubernetesEnvironment directly, since modifying `podData` is / should be impossible (it's only populated from pods and deployments that are in the recipe).

garagatyi replied:
yes, I meant a case described by you. I agree that it is rather a case for the environment object itself.

amisevsk commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/KubernetesEnvironment.java:
```diff
@@ -100,21 +132,72 @@ protected KubernetesEnvironment(
+   * Returns deployments that should be created when environment starts.
+   *
+   * <p>Note: This map <b>should not</b> be changed. If objects in the map need to be changed, see
+   * {@link #getPodsData()}
```
This is tricky, there are kind of 3 levels of modification here.

- `#getPods()` and `#getDeployments()` cannot be modified at all, since you have no way of knowing if you have every deployable object in the workspace
- `#getPodsData()` *can* be modified, in the sense that it's what we'd use for provisioning. You cannot add or delete pods/deployments, but you can modify the objects in the map themselves
- `#addPod()` is the only way to change which deployable objects are in the environment.

I'll add some docs to clear this up.

amisevsk replied:
Added some docs, hopefully it's a little more clear.

amisevsk commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/UniqueNamesProvisioner.java:
```diff
@@ -48,15 +59,38 @@ public void provision(T k8sEnv, RuntimeIdentity identity) throws InfrastructureE
-      final String podName = Names.uniquePodName(podMeta.getName(), workspaceId);
+      putLabel(podMeta, Constants.CHE_ORIGINAL_NAME_LABEL, podMeta.getName());
+      final String podName = Names.uniqueResourceName(podMeta.getName(), workspaceId);
       podMeta.setName(podName);
```
I'm not sure, I believe the behavior should be unchanged. The only difference is in method signatures (use meta instead of pod itself to support deployments) and renaming `#uniquePodName() -> #uniqueResourceName()` since it no longer only modifies pods.

amisevsk commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java:
```diff
@@ -712,39 +718,60 @@ protected void doStartMachine(KubernetesServerResolver serverResolver)
+      final ObjectMeta templateMeta = toCreate.getSpec().getTemplate().getMetadata();
+      storeStartingMachine(createdPod, templateMeta, machineConfigs, serverResolver);
     }
     LOG.debug("Pods creation finished in workspace '{}'", workspaceId);
```
Done.

amisevsk commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java:
```diff
@@ -712,39 +718,60 @@ protected void doStartMachine(KubernetesServerResolver serverResolver)
+    }
+  }
+
+  private void startTracingContainersStartup(PodSpec podSpec, ObjectMeta podMeta) {
```
Done.

amisevsk commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/UniqueNamesProvisioner.java:
```diff
@@ -67,4 +101,51 @@ public void provision(T k8sEnv, RuntimeIdentity identity) throws InfrastructureE
+                env -> {
+                  ConfigMapKeySelector configMap = env.getValueFrom().getConfigMapKeyRef();
+                  String originalName = configMap.getName();
+                  configMap.setName(configMapNameTranslation.get(originalName));
```
Done, and added test cases to `UniqueNamesProvisionerTest.java` to cover this case.

amisevsk commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/KubernetesEnvironmentValidator.java:
```diff
@@ -35,10 +35,10 @@
    */
   public void validate(KubernetesEnvironment env) throws ValidationException {
-    checkArgument(!env.getPods().isEmpty(), "Environment should contain at least 1 pod");
+    checkArgument(!env.getPodsData().isEmpty(), "Environment should contain at least 1 pod");
```
Good catch, fixed.

amisevsk commented:
@ashumilova @benoitf ping for review regarding https://github.com/eclipse/che/pull/12203/commits/90eed2651b02ea4c5981b34b00b532239e50bff2. This commit changes recipe validation a fair bit to support other objects and will be useful further on when adding more objects to recipes support.

The short explanation is: Instead of parsing/validating `IPodItem`s, now we parse/validate `ISupportedListItem = IPodItem | IDeploymentItem | IConfigMapItem`.  This affects other areas of the code, as now instead of getting an `IPodItem` it could be an `IDeploymentItem`, which has to be considered when e.g. detecting machines automatically.

sleshchenko commented on infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/KubernetesEnvironment.java:
```diff
@@ -80,18 +95,35 @@ protected KubernetesEnvironment(
+            e -> {
+              PodTemplateSpec podTemplate = e.getValue().getSpec().getTemplate();
+              podData.put(
+                  e.getKey(), new PodData(podTemplate.getSpec(), podTemplate.getMetadata()));
```
@amisevsk During resolving conflicts with my PR I discovered that
<pre>
PodData {
  "test-deployments": PodData{
                                metadata: 
                                    name: "deployment-pod"
                                    ....
                                spec: ....
                              }
</pre>
It is done by purpose? Would not be better to sync pod name and map key? If it would be better - I can do it in my PR.

sleshchenko replied:
The corresponding unit test are here https://github.com/eclipse/che/pull/12335/files#diff-7b6478093925317e7c597de94e907f45R258

amisevsk replied:
I'm actually not sure which is better -- I don't think keys are really used in a very direct way. IIRC we pretty much always just iterate through the values in the map.

The issues with syncing the `PodData` key with pod name for deployments are
1. We would also need to modify the key for the deployments map, as otherwise using the same key between the two wouldn't work. Also, either the PodData key doesn't match the resource name, or the deployments map key doesn't match the resource name.
2. The actual pod name in a deployment is irrelevant, as pod name is generated from deployment name. This could be confusing in debugging or testing, as this value is basically ignored.

Personally, I'm open to either choice so long as every key in PodData matches some key in the deployments or pods map. I lean slightly towards keeping it the way it is, as this keeps the key in sync with the deployment name, but I'm not firm either way.

One thing that I haven't tested yet but would be interested to know, is if pod name in deployments is even necessary. This could be an edge case for how PodData is handled. I can look into it once I get back to working on this bit of code.
